synapse/lib/stormlib/imap.py

```diff
@@ -0,0 +1,376 @@
+
+                    ($ok, $valu) = $server.markSeen("8182")
+
+                Mark ranges of messages as seen::
+
+                    ($ok, $valu) = $server.markSeen("1:3,6:9")
+            ''',
+            'type': {
+                'type': 'function', '_funcname': 'markSeen',
+                'args': (
+                    {'type': 'str', 'name': 'uid_set',
+                     'desc': 'The UID message set to apply the flag to.'},
+                ),
+                'returns': {
+                    'type': 'list',
+                    'desc': 'An ($ok, $valu) tuple.'
+                },
+            },
+        },
+        {
+            'name': 'markDeleted',
+            'desc': '''
+            Mark messages as deleted by an RFC2060 UID message set.
+
+            The command uses the +FLAGS.SILENT command and applies the \\Deleted flag.
+            By default the expunge command will also be executed on the mailbox.
+            The actual behavior of these commands are mailbox configuration dependent.
+
+            Examples:
+                Mark a single messsage as deleted and expunge::
+
+                    ($ok, $valu) = $server.markDeleted("8182")
+
+                Mark ranges of messages as deleted and expunge::
+
+                    ($ok, $valu) = $server.markDeleted("1:3,6:9")
+
+                Mark a set of messages of messages as deleted but do not expunge::
+
+                    ($ok, $valu) = $server.markDeleted("1,4,6", expunge=$lib.false)
+            ''',
+            'type': {
+                'type': 'function', '_funcname': 'markDeleted',
+                'args': (
+                    {'type': 'str', 'name': 'uid_set',
+                     'desc': 'The UID message set to apply the flag to.'},
+                    {'type': 'boolean', 'name': 'expunge', 'default': True,
+                     'desc': 'Execute the expunge command on the mailbox after setting the flag.'},
+                ),
+                'returns': {
+                    'type': 'list',
+                    'desc': 'An ($ok, $valu) tuple.'
+                },
+            },
+        },
+    )
+    _storm_typename = 'storm:imap:server'
+
+    def __init__(self, runt, imap_cli, path=None):
+        s_stormtypes.StormType.__init__(self, path=path)
+        self.runt = runt
+        self.imap_cli = imap_cli
+        self.locls.update(self.getObjLocals())
+
+    def getObjLocals(self):
+        return {
+            'list': self.list,
+            'fetch': self.fetch,
+            'login': self.login,
+            'search': self.search,
+            'select': self.select,
+            'markSeen': self.markSeen,
+            'markDeleted': self.markDeleted,
+        }
+
+    async def login(self, user, passwd):
+        user = await s_stormtypes.tostr(user)
+        passwd = await s_stormtypes.tostr(passwd)
+
+        coro = self.imap_cli.login(user, passwd)
+        await run_imap_coro(coro)
+
+        return True, None
```

Comment:
so anywhere the `run_imap_coro` is used, it can throw a uncaught exception which then breaks the retn style convention here, making it possible to have something teardown a runtime vs allowing someone to do error handling